src/zarr/core/array.py

```diff
@@ -160,25 +166,48 @@ async def get_array_metadata(
     return metadata_dict
 
 
+TArrayMeta = TypeVar("TArrayMeta", ArrayV2Metadata, ArrayV3Metadata)
+
+
 @dataclass(frozen=True)
-class AsyncArray:
-    metadata: ArrayMetadata
+class AsyncArray(Generic[TArrayMeta]):
```

Comment:
Does this foreclose the opportunity to use other generics here (namely over dtype)?

xref: https://github.com/zarr-developers/zarr-python/pull/2137#discussion_r1735199042

Reply:
that's a good question, I have no idea! It seems like we would want to make the metadata class generic w.r.t dtype, but then have the array class be generic over dtype too... not sure how to solve that